Updated rod-rod, rod-cylinder, self-contact classes, examples, and tests

Ali-7800 opened the pull request:
Second contact module PR mentioned in [Issue](https://github.com/GazzolaLab/PyElastica/issues/113) 
This PR includes:
- Fixed some strings in test_connection.py and test_contact.py which made the tests pass whether the string was identical or not.
- Refactor: _check_order -> _check_order_type
- Added RodRodContact, RodCylinderContact, and RodSelfContact, which are based on ExternalContact and SelfContact from joint.
- Updated the the rod contact and the rigid body contact examples to use the RodRodContact, RodCylinderContact, and RodSelfContact instead of ExternalContact and RodSelfContact
- Added test_contact_call_on_systems() in test_contact.py.
- Updated the test_contact_specific_functions.py,  test_contact_common_functions.py, and  test_contact_wrapper_classes.py to test RodRodContact, RodCylinderContact, and RodSelfContact.
- Added tests for testing new contact wrapper classes _check_order_and_type method.

armantekinalp commented:
@Ali-7800 can you also share the plots or videos for the all examples you have changed in this PR as future reference. Thanks.

Ali-7800 commented:
Plectonemes Case:


https://github.com/GazzolaLab/PyElastica/assets/47090295/418070f8-30e4-41b1-9f56-d08f9bd56e54


https://github.com/GazzolaLab/PyElastica/assets/47090295/a8b9d4a6-9e8d-4310-9bff-afbfbaca1509


https://github.com/GazzolaLab/PyElastica/assets/47090295/e3c8e46d-e290-49f1-9ea0-5e4a09e84f8e


https://github.com/GazzolaLab/PyElastica/assets/47090295/59242456-06b7-4a38-865a-7c8d52fcd56d

![link_twist_writhe](https://github.com/GazzolaLab/PyElastica/assets/47090295/3bb2e2ed-6437-48a2-b173-010517dce6ae)

Solenoids Case:

https://github.com/GazzolaLab/PyElastica/assets/47090295/7032fb76-2a92-4a8b-a29f-4f373cfa6437


https://github.com/GazzolaLab/PyElastica/assets/47090295/4d8437ba-e60e-446a-a809-7950c8947e4c


https://github.com/GazzolaLab/PyElastica/assets/47090295/44826d0e-6398-4604-8f1a-a973473a09db


https://github.com/GazzolaLab/PyElastica/assets/47090295/4c7092ae-ddc8-4423-8abd-c4eb21846114

![link_twist_writhe](https://github.com/GazzolaLab/PyElastica/assets/47090295/957c7b81-6bbc-46ba-b7ad-0ca695d0f461)

rod_rod_contact_inclined_validation:

https://github.com/GazzolaLab/PyElastica/assets/47090295/665ed512-04a4-47ba-aa60-bd9071beb24d


https://github.com/GazzolaLab/PyElastica/assets/47090295/8c4fb038-ed8a-477e-99d3-2f2dbdbbdf19


https://github.com/GazzolaLab/PyElastica/assets/47090295/f595e521-8a07-4937-b975-704ddaf85ee3


https://github.com/GazzolaLab/PyElastica/assets/47090295/80d42cf1-7d26-4b6a-a8ae-52745f896b17

![inclined_rods_velocity](https://github.com/GazzolaLab/PyElastica/assets/47090295/df5d617f-7bd6-4d17-bab5-c35233efeac1)

rod_rod_contact_parallel_validation:

https://github.com/GazzolaLab/PyElastica/assets/47090295/1bac7437-2489-4536-9f53-a64c70e1cd06


https://github.com/GazzolaLab/PyElastica/assets/47090295/2b66fd3a-19db-44f3-a0f6-bfe6461bbdbe


https://github.com/GazzolaLab/PyElastica/assets/47090295/f744dd9d-a542-43bc-93f3-5580da4e0401


https://github.com/GazzolaLab/PyElastica/assets/47090295/6cd0fd14-1c95-4ef7-8edd-a9e631d24cdd

![parallel_rods_velocity](https://github.com/GazzolaLab/PyElastica/assets/47090295/7dea95b0-8661-450e-b092-38d89748daf3)

Rod Cylinder Parallel Contact:

https://github.com/GazzolaLab/PyElastica/assets/47090295/e2d984a2-45f9-4717-895a-ca891aeb2883


https://github.com/GazzolaLab/PyElastica/assets/47090295/d002efda-6422-4b3c-81c0-77117f404140


https://github.com/GazzolaLab/PyElastica/assets/47090295/f6ee1448-e117-4525-a78c-c9de8f9779a3


https://github.com/GazzolaLab/PyElastica/assets/47090295/bcbe2533-3e0b-4fbc-9182-09e9767bf6ac

![rod_rigid_velocity](https://github.com/GazzolaLab/PyElastica/assets/47090295/00f2a0d5-7734-4d82-8c93-de00fe5f6d0c)

Rod Cylinder Inclined Contact:

https://github.com/GazzolaLab/PyElastica/assets/47090295/62edc185-84dd-443a-9872-ad60e19fe6e1


https://github.com/GazzolaLab/PyElastica/assets/47090295/f49aa77d-62fc-4a45-b5f8-91e49cf5deca


https://github.com/GazzolaLab/PyElastica/assets/47090295/29702b4f-1fdd-4b6a-a03e-f9e1b3e083ea


https://github.com/GazzolaLab/PyElastica/assets/47090295/e9c26eaf-bfbd-434b-ba06-e38ca23185de

![rod_rigid_velocity](https://github.com/GazzolaLab/PyElastica/assets/47090295/e799bae1-231a-451e-b314-7487ce94f4f3)

Rod Cylinder Contact with y normal:

https://github.com/GazzolaLab/PyElastica/assets/47090295/3a7eceb8-1314-4453-8cf3-0f53658c1208


https://github.com/GazzolaLab/PyElastica/assets/47090295/6830c37f-2742-496b-9ca9-aacb86d10025

![rod cylinder with y normal](https://github.com/GazzolaLab/PyElastica/assets/47090295/504b83c3-db5e-4c8d-ad07-da92b292a2e2)

Rod Cylinder Contact validation:

https://github.com/GazzolaLab/PyElastica/assets/47090295/ebefa347-1548-4f5a-b751-111b2ff6c1b6


https://github.com/GazzolaLab/PyElastica/assets/47090295/d05d1378-7d03-4025-9a12-3f6934011ba5

![rod cylinder contact validation](https://github.com/GazzolaLab/PyElastica/assets/47090295/f5d6ac64-e43e-41fa-8f68-b597a17f1b58)

Rod Cylinder Contact Friction:

https://github.com/GazzolaLab/PyElastica/assets/47090295/f40275b9-5b8b-463c-84cf-bb0359407426


https://github.com/GazzolaLab/PyElastica/assets/47090295/98c081e4-cece-4e92-a86d-72b909b47a03


https://github.com/GazzolaLab/PyElastica/assets/47090295/095b9196-c09e-4f6f-bf0a-67f713f24025


https://github.com/GazzolaLab/PyElastica/assets/47090295/84180529-2211-4178-a524-9256b7741e6e

![rod_rigid_velocity](https://github.com/GazzolaLab/PyElastica/assets/47090295/30018053-01e5-4ad9-9a96-ee7dbf586366)

Rod Cylinder Contact Friction Phase Space:

![rod_energy_vs_force](https://github.com/GazzolaLab/PyElastica/assets/47090295/bae99dba-dfc2-417f-be19-35ec16e79b1c)

armantekinalp commented on tests/test_modules/test_contact.py:
```diff
@@ -282,20 +300,93 @@ def test_contact_check_order(self, load_contact_objects_with_incorrect_order):
-            RigidBodyBase, RodBase
-        ) in str(
-            excinfo.value
+        assert (
+            "Systems provided to the contact class have incorrect order. \n"
+            " First system is {0} and second system is {1}. \n"
+            " If the first system is a rod, the second system can be a rod, rigid body or surface. \n"
+            " If the first system is a rigid body, the second system can be a rigid body or surface."
+        ).format(mock_rigid_body.__class__, mock_rod.__class__) == str(excinfo.value)
```
Why do you have this assert here? I thought it is checked in the contact function wrapper tests.

Ali-7800 replied:
This is for the NoContact class, the tests in contact function wrapper tests are for the other classes.

armantekinalp commented on tests/test_contact_wrapper_classes.py:
why did you call `test_contact_wrapper_classes.py` instead of `test_contact_classes.py`

Ali-7800 replied:
This is what @Rahul-JOON named it so I used the same file.

armantekinalp replied:
I think we can rename this what do you think @skim0119 @bhosale2

Ali-7800 replied:
renamed

armantekinalp commented on tests/test_contact_specific_functions.py:
@bhosale2 and @skim0119  what about we call this function `test_contact_functions.py`?

Ali-7800 replied:
done

armantekinalp reviewed (changes requested):
Some changes.

Ali-7800 commented on elastica/contact_forces.py:
```diff
@@ -20,7 +26,7 @@ def __init__(self):
         """
 
-    def _check_order(
+    def _check_order_and_type(
```
I was thinking of changing it to `_check_systems_validity`, cause some of them also check whether the rods are the same or distinct which is not really and order or type check. What do you think?

armantekinalp replied:
I like that better. Lets change to that.

skim0119 commented on elastica/contact_forces.py:
```diff
@@ -66,3 +72,832 @@ def apply_contact(
         pass
+
+
+@numba.njit(cache=True)
```
I feel like it is not a right place for these functions to be.

armantekinalp replied:
Yeah they were like this before in joint.py as well. Any suggestions @skim0119 ? We can have a separate file called `contact_utils.py` ? What do you guys think

Ali-7800 replied:
I think this is a good idea, especially when we start adding the new contact functions it will get pretty crowded.

Ali-7800 replied:
moved

armantekinalp commented:
@Ali-7800 can you also check why coverage is going down by 6%?

Ali-7800 commented:
> @Ali-7800 can you also check why coverage is going down by 6%?

it's because I changed the contact tests from using `ExternalContact` and `SelfContact` in joint to the new contact classes. so `ExternalContact` and `SelfContact` in joint are not being tested anymore.

Ali-7800 commented:
> > @Ali-7800 can you also check why coverage is going down by 6%?
> 
> it's because I changed the contact tests from using `ExternalContact` and `SelfContact` in joint to the new contact classes. so `ExternalContact` and `SelfContact` in joint are not being tested anymore.

Added old tests back to in test_contact_specfic_functions.py, test_contact_common_functions.py, and test_contact_wrapper_classes.py so coverage does not go down. Should be removed later with ExternalContact and SelfContact in joint.py

Ali-7800 commented:
Added warning message to ExternalContact and SelfContact in joint.py about deprecation.

armantekinalp commented on elastica/joint.py:
```diff
@@ -954,6 +955,18 @@ def __init__(self, k, nu, velocity_damping_coefficient=0, friction_coefficient=0
         self.velocity_damping_coefficient = velocity_damping_coefficient
         self.friction_coefficient = friction_coefficient
+        log = logging.getLogger()
+        log.warning(
+            # Remove warning and add error if ExternalContact is used in v0.3.3
+            # Remove the option to use ExternalContact, beyond v0.4.0
+            "The option to use the ExternalContact joint for the rod-rod and rod-cylinder "
+            "contact is now deprecated. Instead, for rod-rod contact or rod-cylinder contact, "
+            "use RodRodContact or RodCylinderContact from the add-on Contact mixin class. "
+            "For reference see the classes elastica.contact_forces.RodRodContact() and elastica.contact_forces.RodCylinderContact(),"
+            " and for usage check examples/RigidbodyCases/RodRigidBodyContact/rod_cylinder_contact.py and examples/RodContactCase/RodRodContact/rod_rod_contact_parallel_validation.py."
+            " The option to use the ExternalContact joint for the rod-rod and rod-cylinder"
+            " will be removed in the future (v0.3.3)."
+        )
```
I think all this message will be prompted in one line. Could you please check and divide into couple of lines for readability.

Ali-7800 replied:
<img width="1173" alt="Screenshot 2023-08-20 at 2 26 20 AM" src="https://github.com/GazzolaLab/PyElastica/assets/47090295/aab3f28d-cb76-4838-b41b-36dd60848bac">

armantekinalp commented on elastica/joint.py:
```diff
@@ -1058,6 +1071,18 @@ class SelfContact(FreeJoint):
 
     def __init__(self, k, nu):
         super().__init__(k, nu)
+        log = logging.getLogger()
+        log.warning(
+            # Remove warning and add error if SelfContact is used in v0.3.3
+            # Remove the option to use SelfContact, beyond v0.4.0
+            "The option to use the SelfContact joint for the rod self "
+            "contact is now deprecated. Instead, for rod self contact "
+            "use RodSelfContact from the add-on Contact mixin class."
+            "For reference see the class elastica.contact_forces.RodSelfContact(),"
+            " and for usage check examples/RodContactCase/RodSelfContact/solenoids.py."
+            " The option to use the SelfContact joint for the rod self contact"
+            " will be removed in the future (v0.3.3)."
+        )
 
```
Same as above

Ali-7800 replied:
<img width="1173" alt="Screenshot 2023-08-20 at 2 27 06 AM" src="https://github.com/GazzolaLab/PyElastica/assets/47090295/c05209e6-15f1-4ccb-b787-40098b78abab">

armantekinalp commented on tests/test_contact_utils.py:
```diff
@@ -0,0 +1,367 @@
+__doc__ = """ Test common functions used in contact in Elastica.joint implementation"""
```
Update the docs here.

Ali-7800 replied:
updated

armantekinalp reviewed (changes requested):
Minor changes.

armantekinalp commented on elastica/joint.py:
```diff
@@ -954,6 +955,16 @@ def __init__(self, k, nu, velocity_damping_coefficient=0, friction_coefficient=0
+        log = logging.getLogger()
+        log.warning(
+            # Remove warning and add error if ExternalContact is used in v0.3.3
+            # Remove the option to use ExternalContact, beyond v0.4.0
```
```suggestion
            # Remove the option to use ExternalContact, beyond v0.3.3
```

armantekinalp commented on elastica/joint.py:
```diff
@@ -1058,6 +1069,15 @@ class SelfContact(FreeJoint):
+        log = logging.getLogger()
+        log.warning(
+            # Remove warning and add error if SelfContact is used in v0.3.3
+            # Remove the option to use SelfContact, beyond v0.4.0
```
```suggestion
            # Remove the option to use SelfContact, beyond v0.3.3
```

armantekinalp reviewed (approved):
LGTM two minor suggestions

bhosale2 reviewed (approved):
LGTM

skim0119 commented on elastica/joint.py:
```diff
@@ -954,6 +955,16 @@ def __init__(self, k, nu, velocity_damping_coefficient=0, friction_coefficient=0
         super().__init__(k, nu)
         self.velocity_damping_coefficient = velocity_damping_coefficient
         self.friction_coefficient = friction_coefficient
+        log = logging.getLogger()
```
```suggestion
        log = logging.getLogger(self.__class__.__name__)
```

skim0119 commented on elastica/joint.py:
```diff
@@ -1058,6 +1069,15 @@ class SelfContact(FreeJoint):
 
     def __init__(self, k, nu):
         super().__init__(k, nu)
+        log = logging.getLogger()
```
```suggestion
        log = logging.getLogger(self.__class__.__name__)
```

skim0119 reviewed (approved):
Minor changes. Otherwise, LGTM
